omroneipApp/src/drvOmroneip.cpp

```diff
@@ -0,0 +1,1882 @@
+              continue;
+            }
+            char *pData = (char *)malloc((size_t)(unsigned int)(string_length));
+            status = plc_tag_get_string(x.second->tagIndex, offset, pData, string_length);
+            if (status !=0){
+              asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Tag index: %d Error occured in libplctag while accessing STRING data: %s\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              continue;
+            }
+            status = setStringParam(x.first, pData);
+            asynPrint(pasynUserSelf, ASYN_TRACEIO_DRIVER, "%s:%s My asyn parameter ID: %d My tagIndex: %d My data: %s My type %s\n", driverName, functionName, x.first, x.second->tagIndex, pData, x.second->dataType.c_str());
+            free(pData);
+          }
+          else if (x.second->dataType == "WORD")
+          {
+            int bytes = 2;
+            uint8_t *rawData = (uint8_t *)malloc((size_t)(uint8_t)bytes);
+            status = plc_tag_get_raw_bytes(x.second->tagIndex, offset, rawData, bytes);
+            if (status !=0){
+              asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Tag index: %d Error occured in libplctag while accessing WORD data: %s\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              continue;
+            }
+            epicsInt8 *pData = (epicsInt8 *)malloc(bytes * sizeof(epicsInt8));
+            /* We flip around the hex numbers to match what is done in the PLC */
+            int n = bytes-1;
+            char hexString[bytes*2+1]{};
+            for (int i = 0; i < bytes; i++)
+            {
+              sprintf(hexString+strlen(hexString), "%02X", rawData[n]);
+              pData[i] = rawData[n];
+              n--;
+            }
+            status = doCallbacksInt8Array(pData, bytes, x.first, 0);
+            asynPrint(pasynUserSelf, ASYN_TRACEIO_DRIVER, "%s:%s My asyn parameter ID: %d My tagIndex: %d My data: %s My type %s\n", driverName, functionName, x.first, x.second->tagIndex, hexString, x.second->dataType.c_str());
+            free(rawData);
+            free(pData);
+          }
+          else if (x.second->dataType == "DWORD")
+          {
+            int bytes = 4;
+            uint8_t *rawData = (uint8_t *)malloc((size_t)(uint8_t)bytes);
+            status = plc_tag_get_raw_bytes(x.second->tagIndex, offset, rawData, bytes);
+            if (status !=0){
+              asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Tag index: %d Error occured in libplctag while accessing DWORD data: %s\n", driverName, functionName, x.second->tagIndex, plc_tag_decode_error(status));
+              continue;
+            }
+            epicsInt8 *pData = (epicsInt8 *)malloc(bytes * sizeof(epicsInt8));
+            /* We flip around the hex numbers to match what is done in the PLC */
```

Comment:
In this and the following code sections, you have a lot of malloc and free calls.  It looks like the number of bytes you allocate is often fixed for non array types.  Might it be worth performing single malloc calls outside the loop and then re-using the memory wherever possible.  So for many of the cases you could probably allocate 8 bytes (or similar) and then just use however many bytess you need.
This would perhaps reduce the number of CPU cycles if you had a large database with lots of requests running at high speed.  It would also help to reduce memory fragmentation and make memory tracking easier.
For the arrays, (and this is more complicated), you could also create a storage class to hold memory of varying sizes.  Whenever you need memory ask the storage class if it has already allocated this amount and re-use.  Because I'm assuming once the records have been setup they will always ask for the same qty of items?  This would then definitely reduce memory malloc/frees, in fact eliminate them altogether.  You would simply free everything at the end when you break out of the loop.

Reply:
I like the idea of a wrapper class with RAII - free all the memory when going out of scope.

Reply:
This might be quite involved, so I will create a separate branch and merge request for this later on.